packages/config-helpers/src/define-config.js

```diff
@@ -0,0 +1,380 @@
+			 */
+
+			const entry = [];
+
+			if (Array.isArray(baseFile)) {
+				entry.push(...baseFile);
+			} else {
+				entry.push(baseFile);
+			}
+
+			if (Array.isArray(extensionFile)) {
+				entry.push(...extensionFile);
+			} else {
+				entry.push(extensionFile);
+			}
+
+			result.push(entry);
+		}
+	}
+
+	return result;
+}
+
+/**
+ * Extends a config object with another config object.
+ * @param {Config} baseConfig The base config object.
+ * @param {string} baseConfigName The name of the base config object.
+ * @param {Config} extension The extension config object.
+ * @param {string} extensionName The index of the extension config object.
+ * @return {Config} The extended config object.
+ */
+function extendConfig(baseConfig, baseConfigName, extension, extensionName) {
+	const result = { ...extension };
+
+	// for files we need to create every combination of base and extension files
+	if (baseConfig.files) {
+		result.files = extendConfigFiles(baseConfig.files, extension.files);
+	}
+
+	// for ignores we just concatenation the extension ignores onto the base ignores
+	if (baseConfig.ignores) {
+		result.ignores = baseConfig.ignores.concat(extension.ignores ?? []);
+	}
+
+	result.name = `${baseConfigName} > ${extensionName}`;
+
+	return result;
+}
+
+/**
+ * Processes a list of extends elements.
+ * @param {ConfigWithExtends} config The config object.
+ * @param {WeakMap<Config, string>} configNames The map of config objects to their names.
+ * @return {Config[]} The flattened list of config objects.
+ */
+function processExtends(config, configNames) {
+	if (!config.extends) {
+		return [config];
+	}
+
+	const {
+		/** @type {Config[]} */
+		extends: extendsList,
+
+		/** @type {Config} */
+		...configObject
+	} = config;
+
+	const extensionNames = new WeakMap();
+
+	// replace strings with the actual configs
+	const objectExtends = extendsList.map(extendsElement => {
+		if (typeof extendsElement === "string") {
+			const pluginConfig = findPluginConfig(config, extendsElement);
+
+			// assign names
+			if (Array.isArray(pluginConfig)) {
+				pluginConfig.forEach((pluginConfigElement, index) => {
+					extensionNames.set(
+						pluginConfigElement,
+						`${extendsElement}[${index}]`,
+					);
+				});
+			} else {
+				extensionNames.set(pluginConfig, extendsElement);
+			}
+
+			return pluginConfig;
+		}
+
+		return /** @type {Config} */ (extendsElement);
+	});
+
+	const result = [];
+
+	for (const { indexPath, value: extendsElement } of flatTraverse(
+		objectExtends,
+	)) {
+		const extension = /** @type {Config} */ (extendsElement);
+		const baseConfigName = /** @type {string} */ (configNames.get(config));
+		const extensionName =
+			extensionNames.get(extendsElement) ??
+			getExtensionName(extendsElement, indexPath);
+
+		result.push(
+			extendConfig(
+				configObject,
+				baseConfigName,
+				extension,
+				extensionName,
+			),
+		);
+	}
+
+	// original config last
+	result.push(configObject);
```

Comment:
If the base config has only `ignores` (and possibly `name`) along with `extends`, I think the intent is only to restrict the extended config to files that don't match `ignores`. But, returning the original config in that case additionally creates global ignores which probably isn't the desired outcome.

```js
import { defineConfig } from "@eslint/config-helpers";

console.log(defineConfig([
    {
        ignores: ["tests/**"],
        extends: [{
            rules: {
                "no-unused-vars": "error"
            },
        }],
    }
]));
```

Result:

```js
[
  {
    rules: { 'no-unused-vars': 'error' },
    ignores: [ 'tests/**' ],
    name: 'UserConfig[0][0] > ExtendedConfig[0]'
  },
  { ignores: [ 'tests/**' ] }
]
```

Reply:
> It seems like it would be better to just not include the original config?

Yes, in this case it seems better to just not include the original config.